kythe/cxx/indexer/textproto/recordio_textparser.cc

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2023 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kythe/cxx/indexer/textproto/recordio_textparser.h"
+
+#include <sstream>
+
+#include "absl/functional/function_ref.h"
+#include "absl/strings/ascii.h"
+#include "absl/strings/match.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/strip.h"
+#include "absl/types/optional.h"
+#include "glog/logging.h"
+
+namespace kythe {
+namespace lang_textproto {
+
+namespace {
+
+// WithChar is a delimiter for absl::StrSplit() that splits on given char but
+// also includes the delimiter char.
+struct WithChar {
+  explicit WithChar(char ch) : delimiter_(ch) {}
+  absl::string_view Find(absl::string_view text, size_t pos) const {
+    absl::string_view sep = delimiter_.Find(text, pos);
+    // Always return a zero-width span after the delimiter, so that it's
+    // included if present.
+    sep.remove_prefix(sep.size());
+    return sep;
+  }
+
+ private:
+  absl::ByChar delimiter_;
+};
+
+void ExtendBy(absl::string_view& data, std::size_t count) {
+  data = absl::string_view(data.data(), data.size() + count);
+}
+
+}  // namespace
+
+void ParseRecordTextChunks(
+    absl::string_view content, absl::string_view record_delimiter,
+    absl::FunctionRef<void(absl::string_view chunk, int chunk_start_line)>
+        callback) {
+  int currentline = 0, chunk_begin_line = 0;
+  // Keep track of whether any lines in the chunk contains textproto data
+  // and not all the lines in the chunk are only comment/empty line.
+  bool seen_proto_chunk = false;
+  absl::string_view trimmed_delimiter =
+      absl::StripAsciiWhitespace(record_delimiter);
+  absl::string_view chunk;
+  for (absl::string_view line : absl::StrSplit(content, WithChar('\n'))) {
```

Comment:
Optional: I suspect most of this function can be replaced with StrSplit and a delimiter class and be simpler, e.g.

```
class ProtoLineDelimiter {
 public:
  explicit ProtoLineDelimiter(absl::string_view delimiter, int* line_count = nullptr)
      : delimiter_(delimiter), line_count_(line_count) {}

  /// \brief Finds the next occurrence of the configured delimiter
  /// on a line by itself, after the first non-comment, non-empty line.
  absl::string_view Find(absl::string_view text, size_t pos) {
    for (absl::string_view line :
         absl::StrSplit(text.substr(pos), WithChar('\n'))) {
      if (line_count_) {
        ++*line_count_;
      }
      // Don't look for the delimiter until we've seen our first non-empty,
      // non-comment line.
      data_seen_ = data_seen_ || !(absl::StartsWith(line, "#") ||
                                              absl::StripPrefix(line, "\n").empty());
      if (!data_seen_) continue;

      if (absl::string_view stripped = absl::StripPrefix(line, delimiter_);
          // The line consists entirely of the delimiter.
          data_seen_ && absl::StripPrefix(stripped, "\n").empty()) {
        return line;
      }
    }
    return text.substr(text.size());
  }

 private:
  std::string delimiter_;
  int* line_count_;

  bool data_seen_ = false;
};
...
int line_count = 0;
for (absl::string_view chunk : absl::StrSplit(data, ProtoLineDelimiter(record_delimiter, &line_count)) {
  callback(chunk, line_count);
}
```

This lets StrSplit itself keep track of all of the chunk-information and the loop just has to count lines and find the next delimiter.

Reply:
I am using this code now.. but changed:
- record delimiter could start with a comment.. for example: `# ---`
- each chunk needs the start line of the chunk so made changes to reflect this.